From 193c0c893be5ddc7da9788235b1852a7422a6e85 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:07 -0400
Subject: [ethernet] i40e: Fix dangling ring pointers upon driver removal

Message-id: <1409658042-22689-187-git-send-email-sassmann@redhat.com>
Patchwork-id: 90379
O-Subject: [RHEL7.1 PATCH 186/281] i40e: Fix dangling ring pointers upon driver removal
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

When we resize the number of queues, the driver needs to disassociate any
qvectors that are no longer in use from the original rings, this way we
do not try to access the rings through these qvectors at the time of freeing
the qvectors.

Change-ID: Ie4eb9fc749f8e12348517fe1560f599c58f4a2a4
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 70114ec4aa38096253eef8a9379a59d129e90902)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 3187444..87c538d 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -3151,8 +3151,12 @@ static void i40e_vsi_map_rings_to_vectors(struct i40e_vsi *vsi)
 
  /* If we don't have enough vectors for a 1-to-1 mapping, we'll have to
   * group them so there are multiple queues per vector.
+  * It is also important to go through all the vectors available to be
+  * sure that if we don't use all the vectors, that the remaining vectors
+  * are cleared. This is especially important when decreasing the
+  * number of queues in use.
   */
- for (; v_start < q_vectors && qp_remaining; v_start++) {
+ for (; v_start < q_vectors; v_start++) {
   struct i40e_q_vector *q_vector = vsi->q_vectors[v_start];
 
   num_ringpairs = DIV_ROUND_UP(qp_remaining, q_vectors - v_start);
-- 
1.7.1